sdk/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala

```diff
@@ -36,31 +36,84 @@ import com.digitalasset.daml.lf.transaction.{
 import com.digitalasset.daml.lf.value.Value.ValueArithmeticError
 import com.digitalasset.daml.lf.value.{ContractIdVersion, Value => V}
 
-import scala.annotation.{nowarn, tailrec}
+import scala.annotation.{nowarn, tailrec, unused}
 import scala.collection.immutable.ArraySeq
+import scala.reflect.ClassTag
 import scala.util.control.NonFatal
 
 private[lf] object Speedy {
 
   // These have zero cost when not enabled. But they are not switchable at runtime.
   private val enableInstrumentation: Boolean = false
 
-  final class Metrics(val batchSize: Long) {
+  trait MetricPlugin {
+    type Result
+
+    def incrCount(ctx: MetricPlugin.Ctx*): Unit
+
+    def totalCount: Result
+  }
+  object MetricPlugin {
+    trait Ctx
+  }
+
+  final class StepCount(val batchSize: Long) extends MetricPlugin {
+    type Result = Long
+
     // Speedy evaluates in steps which are grouped into batches of batchSize
     private[this] var stepBatchCount: Long = 0
     private[this] var stepCount: Long = 0
+
+    override def incrCount(ctx: MetricPlugin.Ctx*): Unit = ctx match {
+      case Seq(StepCount.StepCtx) =>
+        stepCount += 1
+      case Seq(StepCount.BatchCtx) =>
+        stepBatchCount += 1
+        stepCount = 0
+      case _ =>
+    }
+
+    override def totalCount: Result = stepBatchCount * batchSize + stepCount
+  }
+  object StepCount {
+    sealed trait Ctx extends MetricPlugin.Ctx
+    case object BatchCtx extends Ctx
+    case object StepCtx extends Ctx
+  }
+
+  final class TxNodeCount extends MetricPlugin {
+    type Result = Long
+
     private[this] var txNodeCount: Long = 0
 
-    private[speedy] def incrStepCount(): Unit = stepCount += 1
-    private[speedy] def incrStepBatchCount(): Unit = {
-      stepBatchCount += 1
-      stepCount = 0
+    override def incrCount(@unused ctx: MetricPlugin.Ctx*): Unit = {
+      txNodeCount += 1
+    }
+
+    override def totalCount: Result = {
+      txNodeCount
     }
-    private[speedy] def incrTransactionNodeCount(): Unit = txNodeCount += 1
+  }
 
-    private[lf] def totalStepCount: (Long, Long) = (stepBatchCount, stepCount)
+  final class Metrics {
+    private[this] var registeredPlugins: Map[String, MetricPlugin] = Map.empty
 
-    private[lf] def transactionNodeCount: Long = txNodeCount
+    private[lf] def register(plugin: MetricPlugin): Unit = {
+      // TODO: info log when plugins get registered
+      registeredPlugins = registeredPlugins.updated(plugin.getClass.getSimpleName, plugin)
+    }
+
+    private[speedy] def incrCount[P <: MetricPlugin: ClassTag](ctx: MetricPlugin.Ctx*): Unit = {
+      registeredPlugins.get(implicitly[ClassTag[P]].runtimeClass.getSimpleName).foreach {
+        _.incrCount(ctx: _*)
+      }
+    }
+
+    private[lf] def totalCount[P <: MetricPlugin: ClassTag]: Option[P#Result] = {
+      registeredPlugins
+        .get(implicitly[ClassTag[P]].runtimeClass.getSimpleName)
+        .map(_.totalCount.asInstanceOf[P#Result])
```

Comment:
Need the `P#Result` as Scala can't perform the type equality against its inferred type